### PR TITLE
Implement the Clone Effect in battle.

### PR DESCRIPTION
**Pull request by Toneyisnow**
1. Make color shifter for CAnimation, and use that in the clone effect.
2. Update the original position of the cloned object.

### CODE REVIEW

File: client/gui/CAnimation.cpp [hunk 1]
```diff
@@ -753,6 +779,24 @@ void SDLImage::shiftPalette(int from, int howMany)
...
+	SDL_Palette* palette = surf->format->palette;
+	for (int i = 0; i < 255; i++)
+	{
+		palette->colors[i] = shifter->shiftColor(originalPalette->colors[i]);
+	}
+}
+
+void SDLImage::resetPalette()
+{
+	SDL_Palette * pal = new SDL_Palette();
```
Comment (alexvins) [inline client/gui/CAnimation.cpp]: `pal` in not used and also leak.

File: client/gui/SDL_Extensions.h [hunk 2]
```diff
@@ -152,6 +152,57 @@ struct ColorPutter
...
+public:
+	virtual SDL_Color shiftColor(SDL_Color clr) = 0;
+};
+
+class ColorShifterLightBlue : public ColorShifter
+{
+public:
+	static ColorShifterLightBlue* create()
+	{
+		return new ColorShifterLightBlue();
```
Comment (alexvins) [inline client/gui/SDL_Extensions.h]: In this case ColorShifter objects will always leak. You could f.e. save local instance with unique_ptr and return pointer to it.

File: client/gui/SDL_Extensions.h [hunk 3]
```diff
@@ -152,6 +152,57 @@ struct ColorPutter
...
+	}
+
+	SDL_Color shiftColor(SDL_Color clr) override
+	{
+		clr.b = clr.b + (255 - clr.b) / 2;
```
Comment (alexvins) [inline client/gui/SDL_Extensions.h]: This method should be `const` (and `ColorShifter` params where it used should be `const` too)

File: client/gui/CAnimation.h [hunk 4]
```diff
@@ -44,6 +45,8 @@ class IImage
 
 	//only indexed bitmaps, 16 colors maximum
 	virtual void shiftPalette(int from, int howMany) = 0;
+	virtual void adjustPalette(ColorShifter * shifter) = 0;
```
Comment (alexvins) [inline client/gui/CAnimation.h]: `ColorShifter * shifter` should be `const`

File: client/battle/CCreatureAnimation.cpp [hunk 5]
```diff
@@ -142,6 +142,19 @@ void CCreatureAnimation::setType(CCreatureAnim::EAnimType type)
 	play();
 }
 
+void CCreatureAnimation::shiftColor(ColorShifter * shifter)
+{
+	if (forward)
```
Comment (alexvins) [inline client/battle/CCreatureAnimation.cpp]: 1. No space after `if` please (also after `for` in other files)
2. Do not use `{}` around single statement

File: client/battle/CCreatureAnimation.h [hunk 6]
```diff
@@ -104,6 +104,9 @@ class CCreatureAnimation : public CIntObject
 	bool incrementFrame(float timePassed);
 	void setBorderColor(SDL_Color palette);
 
+	// tint color effect
+	void shiftColor(ColorShifter* shifter);
```
Comment (alexvins) [inline client/battle/CCreatureAnimation.h]: Put space around `*`

File: client/gui/CAnimation.cpp [hunk 7]
```diff
@@ -736,6 +755,13 @@ void SDLImage::verticalFlip()
 	surf = flipped;
 }
 
+// Keep the original palette, in order to do color switching operation
+void SDLImage::savePalette()
+{
+	originalPalette = new SDL_Palette();
```
Comment (alexvins) [inline client/gui/CAnimation.cpp]: Where does `originalPalette` being deleted? What if it has been saved already? Also you should use `SDL_AllocPalette`

File: client/gui/CAnimation.cpp [hunk 8]
```diff
@@ -736,6 +755,13 @@ void SDLImage::verticalFlip()
 	surf = flipped;
 }
 
+// Keep the original palette, in order to do color switching operation
+void SDLImage::savePalette()
+{
+	originalPalette = new SDL_Palette();
+	memcpy(originalPalette, surf->format->palette, sizeof(SDL_Palette));
```
Comment (alexvins) [inline client/gui/CAnimation.cpp]: Isn`t SDL has a function to copy palettes? I think `SDL_SetPaletteColors` should be used here.

File: client/gui/CAnimation.cpp [hunk 9]
```diff
@@ -753,6 +779,24 @@ void SDLImage::shiftPalette(int from, int howMany)
 	}
 }
 
+void SDLImage::adjustPalette(ColorShifter * shifter)
+{
+	SDL_Palette* palette = surf->format->palette;
+	for (int i = 0; i < 255; i++)
+	{
+		palette->colors[i] = shifter->shiftColor(originalPalette->colors[i]);
```
Comment (alexvins) [inline client/gui/CAnimation.cpp]: You should not alter whole palette. There are a few special color indexes that should not be changed (see CDefFile::CDefFile for reference).

File: client/gui/CAnimation.cpp (hunk 1 above)
Comment (Toneyisnow) [reply client/gui/CAnimation.cpp]: Good point, mem leak is also a problem I was worrying. I think change them into SDL methods and use unique_ptr would resolve the problem.

File: client/gui/CAnimation.h (hunk 4 above)
Comment (Toneyisnow) [reply client/gui/CAnimation.h]: Updated.

File: client/gui/SDL_Extensions.h (hunk 3 above)
Comment (Toneyisnow) [reply client/gui/SDL_Extensions.h]: I think this comment is in wrong place, since it modified the color.

File: client/gui/SDL_Extensions.h (hunk 2 above)
Comment (Toneyisnow) [reply client/gui/SDL_Extensions.h]: Updated with unique_ptr.

File: client/battle/CCreatureAnimation.cpp (hunk 5 above)
Comment (Toneyisnow) [reply client/battle/CCreatureAnimation.cpp]: Updated.

File: client/gui/CAnimation.cpp (hunk 7 above)
Comment (Toneyisnow) [reply client/gui/CAnimation.cpp]: I will check this and update soon.

File: client/gui/CAnimation.cpp (hunk 9 above)
Comment (Toneyisnow) [reply client/gui/CAnimation.cpp]: I will update soon.

File: client/gui/CAnimation.cpp (hunk 8 above)
Comment (Toneyisnow) [reply client/gui/CAnimation.cpp]: Yeah, I'm using SDL_SetPaletteColors in another function, but I didn't update here.

Comment (Toneyisnow) [reply client/gui/CAnimation.cpp]: Updated.

File: client/gui/SDL_Extensions.h [hunk 10]
```diff
@@ -152,6 +152,60 @@ struct ColorPutter
...
+public:
+	virtual SDL_Color shiftColor(SDL_Color clr) const = 0;
+};
+
+class ColorShifterLightBlue : public ColorShifter
+{
+public:
+	static ColorShifterLightBlue * create()
+	{
+		std::unique_ptr<ColorShifterLightBlue> shifter(new ColorShifterLightBlue());
```
Comment (alexvins) [inline client/gui/SDL_Extensions.h]: `shifter` should be static

File: client/gui/SDL_Extensions.h [hunk 11]
```diff
@@ -152,6 +152,60 @@ struct ColorPutter
...
+		return clr;
+	}
+};
+
+class ColorShifterDeepBlue : public ColorShifter
+{
+public:
+	static ColorShifterDeepBlue * create()
+	{
+		std::unique_ptr<ColorShifterDeepBlue> shifter(new ColorShifterDeepBlue());
```
Comment (alexvins) [inline client/gui/SDL_Extensions.h]: same here

File: client/gui/SDL_Extensions.h [hunk 12]
```diff
@@ -152,6 +152,60 @@ struct ColorPutter
...
+		return clr;
+	}
+};
+
+class ColorShifterDeepRed : public ColorShifter
+{
+public:
+	static ColorShifterDeepRed * create()
+	{
+		std::unique_ptr<ColorShifterDeepRed> shifter(new ColorShifterDeepRed());
```
Comment (alexvins) [inline client/gui/SDL_Extensions.h]: and here

File: client/battle/CBattleAnimations.cpp [hunk 13]
```diff
@@ -90,6 +90,12 @@ CBattleStackAnimation::CBattleStackAnimation(CBattleInterface * owner, const CSt
 	assert(myAnim);
 }
 
+void CBattleStackAnimation::shiftColor(ColorShifter * shifter)
```
Comment (alexvins) [inline client/battle/CBattleAnimations.cpp]: `shifter` should be const here too

File: client/battle/CBattleAnimations.h [hunk 14]
```diff
@@ -43,6 +43,8 @@ class CBattleStackAnimation : public CBattleAnimation
 	const CStack * stack; //id of stack whose animation it is
 
 	CBattleStackAnimation(CBattleInterface * _owner, const CStack * _stack);
+
+	void shiftColor(ColorShifter * shifter);
```
Comment (alexvins) [inline client/battle/CBattleAnimations.h]: `shifter` should be const here too

File: client/gui/SDL_Extensions.h (hunk 10 above)
Comment (Toneyisnow) [reply client/gui/SDL_Extensions.h]: I didn't plan to make the "shifter" to be Singeleton, so here it only to create a new one.

Comment (alexvins) [reply client/gui/SDL_Extensions.h]: If "shifter" is not singleton, then you need to return smart pointer from this function or just remove it.

Comment (Toneyisnow) [reply client/gui/SDL_Extensions.h]: Removed the functions.

**Pull request merged**